internal/buffer/search.go

```diff
@@ -147,92 +275,103 @@ func (b *Buffer) FindNext(s string, start, end, from Loc, down bool, useRegex bo
+	if err != nil {
+		return -1, Loc{-1, -1}, err
 	}
 
 	charsEnd := util.CharacterCount(b.LineBytes(end.Y))
-	found := 0
 	var deltas []Delta
 
-	for i := start.Y; i <= end.Y; i++ {
-		l := b.LineBytes(i)
-		charCount := util.CharacterCount(l)
-		if (i == start.Y && start.X > 0) || (i == end.Y && end.X < charCount) {
-			// This replacement code works in general, but it creates a separate
-			// modification for each match. We only use it for the first and last
-			// lines, which may use padded regexps
-
-			from := Loc{0, i}.Clamp(start, end)
-			to := Loc{charCount, i}.Clamp(start, end)
-			matches := b.findAll(search, from, to)
-			found += len(matches)
-
-			for j := len(matches) - 1; j >= 0; j-- {
-				// if we counted upwards, the different deltas would interfere
-				match := matches[j]
-				var newText []byte
-				if captureGroups {
-					newText = search.ReplaceAll(b.Substr(match[0], match[1]), replace)
-				} else {
-					newText = replace
-				}
-				deltas = append(deltas, Delta{newText, match[0], match[1]})
+	n := b.findAllFuncFunc(rgrp, start, end, find, func(match []Loc) {
+		deltas = append(deltas, Delta{repl(match), match[0], match[1]})
+	})
+	b.MultipleReplace(deltas)
+
+	deltaX := util.CharacterCount(b.LineBytes(end.Y)) - charsEnd
+	return n, Loc{end.X + deltaX, end.Y}, nil
+}
+
+// ReplaceAll replaces all matches of the regexp `s` in the given area. The
+// new text is obtained from `template` by replacing each variable with the
+// corresponding submatch as in `Regexp.Expand`. The function returns the
+// number of replacements made, the new end position and any error that
+// occured during regexp compilation
+func (b *Buffer) ReplaceAll(s string, start, end Loc, template []byte) (int, Loc, error) {
+	var replace []byte
+
+	find := func(b *Buffer, r RegexpGroup, start, end Loc) []Loc {
+		return b.findDownFunc(r, start, end, func(re *regexp.Regexp, l []byte) []int {
+			match := re.FindSubmatchIndex(l)
+			if match == nil {
+				return nil
 			}
-		} else {
-			newLine := search.ReplaceAllFunc(l, func(in []byte) []byte {
-				found++
-				var result []byte
-				if captureGroups {
-					match := search.FindSubmatchIndex(in)
-					result = search.Expand(result, replace, in, match)
-				} else {
-					result = replace
-				}
-				return result
-			})
-			deltas = append(deltas, Delta{newLine, Loc{0, i}, Loc{charCount, i}})
-		}
+			replace = re.Expand(nil, template, l, match)
+			return match[:2] // this way match[2:] is not transformed to Loc's
+		})
 	}
 
-	b.MultipleReplace(deltas)
+	return b.replaceAllFuncFunc(s, start, end, find, func(match []Loc) []byte {
+		return replace
+	})
+}
+
+// ReplaceAllLiteral replaces all matches of the regexp `s` with `repl` in
+// the given area. The function returns the number of replacements made, the
+// new end position and any error that occured during regexp compilation
+func (b *Buffer) ReplaceAllLiteral(s string, start, end Loc, repl []byte) (int, Loc, error) {
```

Comment:
Why not pass literal as a boolean argument to `ReplaceAll()`?

Reply:
Sure, I can change that. The reason I had chosen `ReplaceAllLiteral` was to imitate Go's `regexp` API.

Reply:
Heh, didn't realize that. Anyway, we don't need to replicate Go's API precisely, we can define whatever API is more convenient for us to use.

Reply:
What I like about `ReplaceAllLiteral` is that I don't have to remember whether a second argument `true` to `ReplaceAll` means "use as regexp" or "use literally". (I believe that in almost all practical purposes, this argument would be a constant for each invocation of `ReplaceAll`, not some variable whose value is not known in advance.)